# Linear Regression with a Categorical Predictor

## Introduction
Linear regression is a machine learning technique that can be used to model the relationship between a quantitative variable and some other variable(s). Those other variables can be either quantitative (e.g., height or salary) or categorical (e.g., job industry or hair color). However, if we want to include categorical predictors in a linear regression model, we need to treat them a little differently than quantitative variables. This notebook will explore the implementation and interpretation of a single categorical predictor with more than two categories.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats

## The Data
- As an example, we’ll use a dataset from StreetEasy that contains information about housing rentals in New York City. For now, we’ll only focus on two columns of this dataset:

    - **rent:** the rental price of each apartment
    - **borough:** the borough that the apartment is located in, with three possible values ('Manhattan', 'Brooklyn', and 'Queens')

In [9]:
df = pd.read_csv("rentals.csv")
rentals = df[["rent", "borough"]]
rentals.head()

rent    borough
0   2550  Manhattan
1  11500  Manhattan
2   3000     Queens
3   4500  Manhattan
4   4795  Manhattan

## The X Matrix
- To understand how we can fit a regression model with a categorical predictor, it’s useful to walk through what happens when we use ``statsmodels.api.OLS.from_formula()`` to create a model. When we pass a formula to this function (like ``'weight ~ height'`` or ``'rent ~ borough'``), it actually creates a new data set, which we don’t see. This new data set is often referred to as the X matrix, and it is used to fit the model.


- When we use a quantitative predictor, the X matrix looks similar to the original data, but with an additional column of `1`s in front (the reasoning behind this column of `1`s is the subject of a future article — for now, no need to worry about it!). However, when we fit the model with a categorical predictor, something else happens: we end up with additional column(s) of `1`s and `0`s.


- For example, let’s say we want to fit a regression predicting `rent` based on `borough`. We can see the X matrix for this model using ``patsy.dmatrices()``, which is implemented behind the scenes in `statsmodels`:

In [10]:
import patsy

In [13]:
y, X = patsy.dmatrices("rent ~ borough", rentals)
X[0:5]

array([[1., 1., 0.],
       [1., 1., 0.],
       [1., 0., 1.],
       [1., 1., 0.],
       [1., 1., 0.]])

- The first column is all `1`s, just like we would get for a quantitative predictor; but the second two columns were formed based on the `borough` variable. Remember that the first five values of the `borough` column looked like this:

In [17]:
rentals.head()

rent    borough
0   2550  Manhattan
1  11500  Manhattan
2   3000     Queens
3   4500  Manhattan
4   4795  Manhattan

- Note that the second column of the X matrix ``[1, 1, 0, 1, 1]`` is an indicator variable for `Manhattan`: it is equal to `1` where the value of borough is ``'Manhattan'`` and `0` otherwise. Meanwhile, the third column of the X matrix (``[0, 0, 0, 1, 1]``) is an indicator variable for Queens: it is equal to `1` where the value of borough is ``'Queens'`` and `0` otherwise.


- The X matrix does not contain an indicator variable for Brooklyn. That’s because this data set only contains three possible values of borough: 'Brooklyn', 'Manhattan', and 'Queens'. In order to recreate the borough column, we only need two indicator columns — because any apartment that is not in 'Manhattan' or 'Queens' must be 'Brooklyn'. For example, the first row of the X matrix has 0s in both indicator columns, indicating that the apartment must be in Brooklyn. Mathematically, we say that a 'Brooklyn' indicator creates collinearity in the X matrix. In regular English: a 'Brooklyn' indicator does not add any new information.


- Because 'Brooklyn' is missing from the X matrix, it is the reference category for this model.

## Implementation and Interpretation

In [18]:
import statsmodels.api as sm

In [21]:
model = sm.OLS.from_formula("rent ~ borough", data = rentals)
model = model.fit()

In [22]:
model.params

Intercept               3327.403751
borough[T.Manhattan]    1811.536627
borough[T.Queens]       -811.256430
dtype: float64

- In the output, we see two different slopes: one for `borough[T.Manhattan]` and one for `borough[T.Queens]`, which are the two indicator variables we saw in the X matrix. We can use the intercept and two slopes to construct the following equation to predict rent:


    - General Equation = rent = 3327.4 + 1811.5 * borough[T.Manhattan] - 811.3 * borough[T.Queens]
    
    - Equaiton 1 : Brooklyn
        - rent = 3327.4 + 1811.5 * 0 - 811.3 * 0
        - rent = 3327.4
        
    - Equation 2 : Manhattan
        - rent = 3327.4 + 1811.5 * 1 - 811.3 * 0
        - rent = 5138.9
       
    - Equation 3 : Queens
        - rent = 3327.4 + 1811.5 * 0 - 811.3 * 1
        - rent = 2516.1
 
- We can verify our understanding of all these coefficients by printing out the average rental prices by borough:

In [23]:
rentals.groupby("borough").mean()

rent
borough               
Brooklyn   3327.403751
Manhattan  5138.940379
Queens     2516.147321

## Changing the Reference Category
- In the example above, we saw that `'Brooklyn'` was the default reference category (because it comes first alphabetically), but we can easily change the reference category in the model as follows:

In [24]:
model = sm.OLS.from_formula("rent ~ C(borough, Treatment('Manhattan'))", rentals)
model = model.fit()

In [25]:
model.params

Intercept                                         5138.940379
C(borough, Treatment('Manhattan'))[T.Brooklyn]   -1811.536627
C(borough, Treatment('Manhattan'))[T.Queens]     -2622.793057
dtype: float64

- In this example, the reference category is 'Manhattan'. Therefore, the intercept is the mean rental price in Manhattan, and the other slopes are the mean differences for Brooklyn and Queens in comparison to Manhattan.

## Other Python Libraries for fitting Linear Models
- There are a few different Python libraries that can be used to fit linear regression models. It is therefore important to understand how this implementation differs for each library. In `statsmodels`, the creation of the X matrix happens completely “behind the scenes” once we pass in a model formula.


- In `scikit-learn` (another popular library for linear regression), we actually need to construct the indicator variables ourselves. Note that we do not have to construct the extra column of `1`s that we saw in the X matrix — this also happens behind the scenes in `scikit-learn`. In order to construct those indicator variables, the `pandas get_dummies()` function is extremely useful:

In [26]:
rentals = pd.get_dummies(data = rentals, columns = ["borough"], drop_first = True)
rentals.head()

rent  borough_Manhattan  borough_Queens
0   2550                  1               0
1  11500                  1               0
2   3000                  0               1
3   4500                  1               0
4   4795                  1               0

- Setting `drop_first = True` tells Python to drop the first indicator variable (for `'Brooklyn'` in thie case), which is what we need for linear regression. We can then fit the exact same model using `scikit-learn` as follows:

In [27]:
from sklearn.linear_model import LinearRegression

In [28]:
X = rentals.drop(columns = "rent")
y = rentals["rent"]

In [30]:
linear_model = LinearRegression()
linear_model.fit(X, y)

LinearRegression()

In [31]:
linear_model.intercept_

3327.4037512339555

In [32]:
linear_model.coef_

array([1811.5366274 , -811.25642981])

## Conclusion
In this article, we’ve walked through an example of how to implement and interpret categorical predictors in a linear regression model. In the process, we’ve learned a little bit about what happens behind the scenes when we fit a linear model using `statsmodels` or `scikit-learn`. This knowledge will help prepare us to fit and interpret more complex models that build upon these foundations.

---